File: 13/main.py
# Juan Manuel González Ascencio
# Leetcode #13 from https://leetcode.com/problems/roman-to-integer/

class Solution:
    def romanToInt(self, s: str) -> int:
        # Dictionary with the values of each roman numeral
        roman = {"I": 1, "V": 5, "X":10, "L": 50, "C":100, "D":500, "M":1000}

        res = 0
        for i in range(len(s)):
            #si el numero actual es menor que el siguiente se resta
            if (i+1 < len(s) and roman[s[i]] < roman[s[i+1]]):
                res -= roman[s[i]]
            #si el numero actual es mayor o igual que el siguiente se suma
            else:
                res += roman[s[i]]

        return res
    

File: 13/main.cpp
//Juan Manuel González Ascencio
//Leetcode #13 from https://leetcode.com/problems/roman-to-integer/
// comando para compilar: g++ main.cpp -o ./app

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

    int values(char c){
        switch(c){
            case 'I': return 1; break;
            case 'V': return 5; break;
            case 'X': return 10; break;
            case 'L': return 50; break;
            case 'C': return 100; break;
            case 'D': return 500; break;
            case 'M': return 1000; break;
        }
        return -1;
    }

    int Roman2Int(string s){
        int res = 0;
        for (int i=0; i<s.length(); i++){
            int s1 = values(s[i]);


            if (i+1 < s.length()){ //if there is a next character
                int s2 = values(s[i+1]); //get the value of the next character
                if (s1 >= s2){ //if the value of the current character is greater than the next one
                    res += s1; //add the value of the current character
                }
                else{
                    res += s2 - s1; //if the value of the current character is less than the next one, add the difference
                    i++;
                }
            }
            else{
                res += s1;
            }
        }
        return res;
    }


};


int main(){
    Solution sol;
    string s = "MCMXCIV";
    cout << sol.Roman2Int(s) << endl;
    return 0;
}